File: banana-video.py
import argparse
import cv2
import os
import sys
from PIL import Image
import numpy as np
import time

ASCII_CHARS = r"$@B%8&WM#*oahkbdpqwmZ0QLCJUYXzcvunxrjft/\|()1{}[]?-_+~<>i!lI;:,\"^`'. "

def frame_to_ascii(frame, width=None, height=None, scale=None, invert=False):
    image = Image.fromarray(frame).convert("L")

    orig_w, orig_h = image.size
    if width and height:
        new_w, new_h = width, height
    elif width:
        new_w = width
        new_h = int((width / orig_w) * orig_h * 0.5)
    elif height:
        new_h = height
        new_w = int((height / orig_h) * orig_w * 2)
    elif scale:
        new_w = int(orig_w * scale)
        new_h = int(orig_h * scale * 0.5)
    else:
        new_w, new_h = orig_w, int(orig_h * 0.5)

    image = image.resize((new_w, new_h))

    ascii_str = ""
    for y in range(new_h):
        for x in range(new_w):
            pixel = image.getpixel((x, y))
            if invert:
                pixel = 255 - pixel
            ascii_str += ASCII_CHARS[pixel * len(ASCII_CHARS) // 256]
        ascii_str += "\n"
    return ascii_str

def main():
    print("banana: convert video to ASCII art")

    parser = argparse.ArgumentParser(
        usage="banana-video.py [-f FILE] [-s SCALE] [-w WIDTH] [-H HEIGHT] [-o OUTPUT] [-t] [-i] [-p] [-frame N -frame-out FILE]",
        description="banana: convert video to ASCII art"
    )

    parser.add_argument('-f', help="Input video file", required=True)
    parser.add_argument('-s', type=float, help="Scale factor")
    parser.add_argument('-w', type=int, help="Output width")
    parser.add_argument('-H', type=int, help="Output height")
    parser.add_argument('-o', help="Output .txt file")
    parser.add_argument('-t', action='store_true', help="Suppress terminal output")
    parser.add_argument('-i', action='store_true', help="Invert brightness")
    parser.add_argument('-p', action='store_true', help="Play video in terminal (override previous frame)")
    parser.add_argument('-frame', type=int, help="Frame number to export as image")
    parser.add_argument('-frame-out', help="Output image file for frame export")

    if len(sys.argv) == 1:
        parser.print_help()
        sys.exit(0)

    args = parser.parse_args()

    if not os.path.isfile(args.f):
        print(f"Error: Video file '{args.f}' does not exist.")
        sys.exit(1)

    cap = cv2.VideoCapture(args.f)
    if not cap.isOpened():
        print("Error: Could not open video.")
        sys.exit(1)

    # Export frame as image if requested
    if args.frame is not None and args.frame_out:
        total = int(cap.get(cv2.CAP_PROP_FRAME_COUNT))
        if args.frame >= total or args.frame < 0:
            print(f"Error: Frame number out of range. Video has {total} frames.")
            sys.exit(1)
        cap.set(cv2.CAP_PROP_POS_FRAMES, args.frame)
        ret, frame = cap.read()
        if not ret:
            print("Error: Could not read the requested frame.")
            sys.exit(1)
        cv2.imwrite(args.frame_out, frame)
        print(f"Frame {args.frame} saved to {args.frame_out}")
        sys.exit(0)

    # In case -p is used, we need to play the video frames in the terminal
    if args.p:
        if args.o:
            print("Error: -p and -o cannot be used together. Please choose one.")
            sys.exit(1)

        frame_count = 0
        while True:
            ret, frame = cap.read()
            if not ret:
                break

            frame = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
            ascii_frame = frame_to_ascii(
                frame,
                width=args.w,
                height=args.H,
                scale=args.s,
                invert=args.i
            )

            # Clear terminal and print the frame
            print("\033c", end="")
            print(ascii_frame, end="")
            time.sleep(0.1)  # Slow down for visual effect (adjust as necessary)

            frame_count += 1

        cap.release()
        return

    frame_count = 0
    output = ""

    while True:
        ret, frame = cap.read()
        if not ret:
            break

        frame = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
        ascii_frame = frame_to_ascii(
            frame,
            width=args.w,
            height=args.H,
            scale=args.s,
            invert=args.i
        )

        if args.o:
            # If -o is specified, write output without frame number header and separate with blank line
            output += f"{ascii_frame}\n"

        if not args.t and not args.p:
            print(f"Frame {frame_count}:")
            print(ascii_frame)
            print("\n")

        frame_count += 1

    cap.release()

    if args.o:
        try:
            with open(args.o, 'w') as f:
                f.write(output)
            print(f"ASCII video written to {args.o}")
        except Exception as e:
            print(f"Error writing to file: {e}")
            sys.exit(1)

if __name__ == "__main__":
    main()

File: banana-video.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
#include <fstream>
#include <string>
#include <thread>
#include <chrono>
#include <getopt.h>

const std::string ASCII_CHARS = R"($@B%8&WM#*oahkbdpqwmZ0QLCJUYXzcvunxrjft/\|()1{}[]?-_+~<>i!lI;:,"^`'. )";

std::string frameToAscii(const cv::Mat& frame, int width = 0, int height = 0, float scale = 0.0, bool invert = false) {
    cv::Mat gray;
    cv::cvtColor(frame, gray, cv::COLOR_BGR2GRAY);

    int orig_w = gray.cols;
    int orig_h = gray.rows;

    int new_w = orig_w, new_h = orig_h * 0.5;

    if (width > 0 && height > 0) {
        new_w = width;
        new_h = height;
    } else if (width > 0) {
        new_w = width;
        new_h = (float(width) / orig_w) * orig_h * 0.5;
    } else if (height > 0) {
        new_h = height;
        new_w = (float(height) / orig_h) * orig_w * 2;
    } else if (scale > 0.0f) {
        new_w = orig_w * scale;
        new_h = orig_h * scale * 0.5;
    }

    cv::resize(gray, gray, cv::Size(new_w, new_h));

    std::string ascii;
    for (int y = 0; y < gray.rows; y++) {
        for (int x = 0; x < gray.cols; x++) {
            int pixel = gray.at<uchar>(y, x);
            if (invert) pixel = 255 - pixel;
            char c = ASCII_CHARS[pixel * ASCII_CHARS.length() / 256];
            ascii += c;
        }
        ascii += '\n';
    }
    return ascii;
}

void printHelp() {
    std::cout << "Usage: banana-video -f FILE [-w WIDTH -H HEIGHT | -s SCALE] [-i] [-o OUTPUT] [-t] [-p] [-frame N -frame-out FILE]\n";
}

int main(int argc, char** argv) {
    std::string file, outputFile, frameOutFile;
    int width = 0, height = 0, frameNum = -1;
    float scale = 0.0;
    bool invert = false, suppress = false, play = false;

    const struct option long_opts[] = {
        {"file", required_argument, nullptr, 'f'},
        {"scale", required_argument, nullptr, 's'},
        {"width", required_argument, nullptr, 'w'},
        {"height", required_argument, nullptr, 'H'},
        {"output", required_argument, nullptr, 'o'},
        {"invert", no_argument, nullptr, 'i'},
        {"suppress", no_argument, nullptr, 't'},
        {"play", no_argument, nullptr, 'p'},
        {"frame", required_argument, nullptr, 1},
        {"frame-out", required_argument, nullptr, 2},
        {nullptr, 0, nullptr, 0}
    };

    int opt;
    while ((opt = getopt_long(argc, argv, "f:s:w:H:o:itp", long_opts, nullptr)) != -1) {
        switch (opt) {
            case 'f': file = optarg; break;
            case 's': scale = std::stof(optarg); break;
            case 'w': width = std::stoi(optarg); break;
            case 'H': height = std::stoi(optarg); break;
            case 'o': outputFile = optarg; break;
            case 'i': invert = true; break;
            case 't': suppress = true; break;
            case 'p': play = true; break;
            case 1: frameNum = std::stoi(optarg); break;
            case 2: frameOutFile = optarg; break;
            default: printHelp(); return 1;
        }
    }

    if (file.empty()) {
        std::cerr << "Error: Missing input file.\n";
        printHelp();
        return 1;
    }

    if ((width || height) && scale > 0.0f) {
        std::cerr << "Error: Use either width/height or scale, not both.\n";
        return 1;
    }

    if ((width && !height) || (!width && height)) {
        std::cerr << "Error: Both width and height must be specified.\n";
        return 1;
    }

    cv::VideoCapture cap(file);
    if (!cap.isOpened()) {
        std::cerr << "Error: Could not open video.\n";
        return 1;
    }

    int totalFrames = cap.get(cv::CAP_PROP_FRAME_COUNT);

    // Frame export
    if (frameNum >= 0 && !frameOutFile.empty()) {
        if (frameNum >= totalFrames || frameNum < 0) {
            std::cerr << "Error: Frame out of range (0 to " << totalFrames - 1 << ").\n";
            return 1;
        }
        cap.set(cv::CAP_PROP_POS_FRAMES, frameNum);
        cv::Mat frame;
        if (cap.read(frame)) {
            cv::imwrite(frameOutFile, frame);
            std::cout << "Saved frame " << frameNum << " to " << frameOutFile << "\n";
            return 0;
        } else {
            std::cerr << "Error: Could not read the requested frame.\n";
            return 1;
        }
    }

    std::ostringstream output;

    int frameIdx = 0;
    while (true) {
        cv::Mat frame;
        if (!cap.read(frame)) break;

        std::string ascii = frameToAscii(frame, width, height, scale, invert);

        if (play) {
            std::cout << "\033[2J\033[H";  // Clear screen
            std::cout << ascii;
            std::this_thread::sleep_for(std::chrono::milliseconds(100));
        } else if (!suppress) {
            std::cout << "Frame " << frameIdx++ << ":\n";
            std::cout << ascii << "\n";
        }

        if (!outputFile.empty()) {
            output << ascii << "\n";
        }
    }

    cap.release();

    if (!outputFile.empty()) {
        std::ofstream out(outputFile);
        if (out) {
            out << output.str();
            std::cout << "Saved output to " << outputFile << "\n";
        } else {
            std::cerr << "Error: Could not write to file.\n";
        }
    }

    return 0;
}
